packages/dashboard/e2e/tests/utils/grid.ts

```diff
@@ -0,0 +1,134 @@
+      const gutter = cellSize / 2;
+      return {
+        x: x + (col * cellSize + gutter),
+        y: y + (row * cellSize + gutter),
+      };
+    },
+    /**
+     *
+     * @returns all of the widget Locators in the grid
+     */
+    widgets: async (): Promise<Locator[]> => await page.locator(WidgetSelector).all(),
+    /**
+     *
+     * @param widgetType which widget to drag and drop from the component pallette
+     * @param offset position offset from the top left corner of grid
+     * @returns the locator of the added widget
+     */
+    addWidget: async (
+      widgetType: ComponentLibrarySelector,
+      offset?: (gridBounds: BoundingBox) => { x: number; y: number }
+    ): Promise<Locator> => {
+      const defaultGridTargetOffset = ({ x, y }: BoundingBox) => ({ x, y }); // Default is 0, 0 on the grid
+      const gridTargetOffset = offset || defaultGridTargetOffset;
+      const widget = page.getByRole('button', { name: WidgetSelectorMap[widgetType] });
+      const draggableWidget = dragGenerator(widget);
+      await draggableWidget.dragTo(gridArea, { targetPosition: ({ target }) => gridTargetOffset(target) });
+      const widgets = await page.locator(WidgetSelector).all();
+      // Assumption that the widget added is last in the DOM
+      const addedWidget = last(widgets);
+      if (!addedWidget) throw new Error('No added widget');
+      return addedWidget;
+    },
+    /**
+     * click the upper right corner of the grid to click out of the selection
+     */
+    clearSelection: async () => {
+      const bounds = await getBoundingBox(gridArea);
+      const { x, y, width } = bounds;
+      await page.mouse.move(x + width - 10, y + 10);
+      await page.mouse.down({ button: 'left' });
+      await page.mouse.up({ button: 'left' });
+    },
+    /**
+     *
+     * @param widgetLocator click the center of the locator
+     */
+    clickWidget: async (widgetLocator: Locator) => {
+      const bounds = await getBoundingBox(widgetLocator);
+      await page.mouse.move(...center(bounds));
+      await page.mouse.down({ button: 'left' });
+      await page.mouse.up({ button: 'left' });
+    },
+    /**
+     * drag select from the top left to the bottom right of the grid
+     */
+    selectAll: async () => {
+      await dragGenerator(gridArea).dragTo(gridArea, {
+        sourcePosition: ({ source }) => ({ x: source.x, y: source.y }),
+        targetPosition: ({ target }) => ({ x: target.x + target.width, y: target.y + target.height }),
+      });
+    },
+    /**
+     *
+     * @param anchor anchor of the selection box to click drag
+     * @param targetPosition the position offset to move the anchor to
+     */
+    resizeSelection: async (anchor: SelectionAnchor, targetPosition: DragPosition) => {
+      const anchorLocator = page.locator(SelectionAnchorMap[anchor]);
+      await dragGenerator(anchorLocator).dragTo(gridArea, {
+        targetPosition,
+      });
+    },
+    /**
+     *
```

Comment:
Why the empty line in this and other doc comments?

Reply:
cleaned up